library/std/src/sys/unix/fs/dir_fd.rs

```diff
@@ -0,0 +1,388 @@
+        // the directory name, inode pair currently being processed
+        let mut current_path_component = PathComponent::from_name_and_fd(
+            unsafe { CStr::from_bytes_with_nul_unchecked(b"\0") },
+            current_readdir.as_fd(),
+        )?;
+        let root_parent_component = PathComponent::from_name_and_fd(
+            unsafe { CStr::from_bytes_with_nul_unchecked(b"\0") },
+            current_readdir.get_parent()?.as_fd(),
+        )?;
+        loop {
+            while let Some(child) = current_readdir.next() {
+                let child = child?;
+                let child_name = child.name_cstr();
+                if let Some(false) = is_dir(&child) {
+                    // just unlink files
+                    cvt(unsafe {
+                        unlinkat(current_readdir.as_fd().as_raw_fd(), child_name.as_ptr(), 0)
+                    })?;
+                } else {
+                    if let Some(child_readdir) =
+                        current_readdir.open_subdir_or_unlink_non_dir(child_name)?
+                    {
+                        // descend into this child directory
+
+                        let child_path_compoment =
+                            PathComponent::from_name_and_fd(child_name, child_readdir.as_fd())?;
+                        path_components.push(current_path_component);
+                        current_path_component = child_path_compoment;
+
+                        // avoid growing the cache over capacity
+                        if readdir_cache.len() == readdir_cache.capacity() {
+                            readdir_cache.pop_front();
+                        }
+                        readdir_cache.push_back(current_readdir);
+                        current_readdir = child_readdir;
+                    }
+                }
+            }
+
+            match path_components.pop() {
+                Some(parent_component) => {
+                    // going back up...
+
+                    // get parent directory readdir
+                    let parent_readdir = match readdir_cache.pop_back() {
+                        Some(readdir) => readdir,
+                        None => {
+                            // cache is empty - reopen parent
+                            let parent_readdir = current_readdir.get_parent()?;
+                            parent_component.verify_dev_ino(parent_readdir.as_fd())?;
+                            parent_readdir
+                        }
+                    };
+
+                    // If `parent_readdir` was now or previously opened via `get_parent()`
+                    // we need to verify the grandparent (dev, inode) pair as well to protect
+                    // against the situation that the child directory was moved somewhere
+                    // else and that the parent just happens to have the same reused
+                    // (dev, inode) pair, that we found descending.
```

Comment:
Its not clear to me how ascending to the grandparent is needed. Or perhaps its just a wording confusion problem.

https://github.com/rust-lang/rust/pull/93160#discussion_r790290684 describes causing a privileged process to delete the contents of a sticky directory by moving.

1. `foo/bar/baz` is being deleted
2. Alice moves `bar` aside
3. Alice deletes `foo`
4. Alice waits for the inode `foo` had to be recycled into a desirable context, for instance as a new sticky dir `/somewhere/newtmp`
5. Alice moves `bar` into that dir
6. the priviled process moves up from `bar` to `/somewhere/newtmp`

At this point the *parent* of `/somewhere/newtmp` needs to be verified as being legitimate, so its (dev,inode) are cross-checked; the grandparent is not relevant.

But as I say perhaps I'm just misunderstanding :/. It seems to me that a rule of 'only operate (readdir/unlink_at its children) on a directory that was either a) opened going down or b) was reopened going up and still has its own original parent (inode,dev) seems quite robust: the greatest attack possible is to operate on newly created directories that have the same inode as they had before *and* are in the same containing directory-by-inode that they had before. That property holds because downward traversal to discover a directory inode must occur before inode reuse can be attempted.

On the other hand, if reuse attacks are something worth guarding against, I don't think that that property is sufficient: the privileged attacker *can* be fooled into deleting content that was not in the original file tree, and if that acts as a gadget into some other attack, we now have a chained attack possible.